src/Microsoft.Health.Dicom.Blob/Features/Storage/BlobFileStore.cs

```diff
@@ -75,6 +78,79 @@ public class BlobFileStore : IFileStore
         }
     }
 
+    /// <inheritdoc />
+    public async Task<Uri> StoreFileInBlocksAsync(
+        long version,
+        Stream stream,
+        IDictionary<string, long> blockLengths,
+        CancellationToken cancellationToken)
+    {
+        EnsureArg.IsNotNull(stream, nameof(stream));
+        EnsureArg.IsNotNull(blockLengths, nameof(blockLengths));
+        EnsureArg.IsGte(blockLengths.Count, 0, nameof(blockLengths.Count));
+
+        BlockBlobClient blobClient = GetInstanceBlockBlobClient(version);
+
+        int maxBufferSize = (int)blockLengths.Max(x => x.Value);
+
+        await ExecuteAsync(async () =>
+        {
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(maxBufferSize);
+            try
+            {
+                foreach (string blockId in blockLengths.Keys)
+                {
+                    long blockSize = blockLengths[blockId];
+
+                    await stream.ReadAsync(buffer.AsMemory(0, (int)blockSize), cancellationToken);
+
+                    using var blockStream = new MemoryStream(buffer);
+                    await blobClient.StageBlockAsync(blockId, blockStream, cancellationToken: cancellationToken);
+
+                    Array.Clear(buffer);
+                }
+
+                await blobClient.CommitBlockListAsync(blockLengths.Keys, cancellationToken: cancellationToken);
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(buffer, clearArray: true);
+            }
+        });
+
+        return blobClient.Uri;
+    }
+
+    /// <inheritdoc />
+    public async Task UpdateFileBlockAsync(
+        long version,
+        string blockId,
+        Stream stream,
+        CancellationToken cancellationToken)
+    {
+        EnsureArg.IsNotNull(stream, nameof(stream));
+        EnsureArg.IsNotNullOrWhiteSpace(blockId, nameof(blockId));
+
+        BlockBlobClient blobClient = GetInstanceBlockBlobClient(version);
+
+        BlockList blockList = await blobClient.GetBlockListAsync(BlockListTypes.Committed, snapshot: null, conditions: null, cancellationToken);
+
+        IEnumerable<string> blockIds = blockList.CommittedBlocks.Select(x => x.Name);
+
+        string blockToUpdate = blockIds.FirstOrDefault(x => x.Equals(blockId, StringComparison.OrdinalIgnoreCase));
+
+        if (blockToUpdate == null)
+            throw new DataStoreException(DicomBlobResource.BlockNotFound);
+
+        stream.Seek(0, SeekOrigin.Begin);
+
+        await ExecuteAsync(async () =>
+        {
+            await blobClient.StageBlockAsync(blockId, stream, cancellationToken: cancellationToken);
+            await blobClient.CommitBlockListAsync(blockIds, cancellationToken: cancellationToken);
```

Comment:
Yeah, we have to commit all the old blockids, else it will overwrite with one block. Strange API.

```
The [StageBlockAsync(String, Stream, BlockBlobStageBlockOptions, CancellationToken)](https://learn.microsoft.com/en-us/dotnet/api/azure.storage.blobs.specialized.blockblobclient.stageblockasync?view=azure-dotnet#azure-storage-blobs-specialized-blockblobclient-stageblockasync(system-string-system-io-stream-azure-storage-blobs-models-blockblobstageblockoptions-system-threading-cancellationtoken)) operation creates a new block as part of a block blob's "staging area" to be eventually committed via the [CommitBlockListAsync(IEnumerable<String>, CommitBlockListOptions, CancellationToken)](https://learn.microsoft.com/en-us/dotnet/api/azure.storage.blobs.specialized.blockblobclient.commitblocklistasync?view=azure-dotnet#azure-storage-blobs-specialized-blockblobclient-commitblocklistasync(system-collections-generic-ienumerable((system-string))-azure-storage-blobs-models-commitblocklistoptions-system-threading-cancellationtoken)) operation.
```